### PR TITLE
fix(geometry): make bbox_to_mask3d preserve the input dtype (#4250)

### LINKED ISSUE #4250: W1b: `bbox_to_mask3d` returns float32 `(B, 1, D, H, W)` while `bbox_to_mask` and `Boxes3D.to_mask` keep the input dtype with no channel axis
The three mask builders already differ in argument order (#4014) and rounding policy (#4015). They also disagree on the output dtype and rank: `bbox_to_mask3d` ends with `.float()` and keeps a singleton channel axis from its intermediate broadcast, so it returns `float32` `(B, 1, D, H, W)` for any input dtype (float64, float16, bfloat16, int64 all executed), while `bbox_to_mask` returns `(B, H, W)` in the input dtype (int64 in gives int64 out) and `Boxes3D.to_mask` returns `(N, D, H, W)` in the box dtype.

Reproduced 2026-09-05 at `1eea5835`, PyTorch 2.14.0, CPU and MPS.

## Reproduction

```python
import torch
from kornia.geometry.bbox import bbox_to_mask, bbox_to_mask3d
from kornia.geometry.boxes import Boxes3D

cuboid = torch.tensor([[[1., 1., 1.], [3., 1., 1.], [3., 3., 1.], [1., 3., 1.],
                        [1., 1., 2.], [3., 1., 2.], [3., 3., 2.], [1., 3., 2.]]], dtype=torch.float64)
square = torch.tensor([[[1., 1.], [3., 1.], [3., 2.], [1., 2.]]], dtype=torch.float64)
for label, out in (("bbox_to_mask3d", bbox_to_mask3d(cuboid, (4, 5, 6))),
                   ("Boxes3D.to_mask", Boxes3D(cuboid).to_mask(4, 5, 6)),
                   ("bbox_to_mask", bbox_to_mask(square, 6, 5))):
    print(label, tuple(out.shape), out.dtype)
```

## Observed

```text
bbox_to_mask3d (1, 1, 4, 5, 6) torch.float32
Boxes3D.to_mask (1, 4, 5, 6) torch.float64
bbox_to_mask (1, 5, 6) torch.float64
```

## Expected

One output contract across the family: the mask in the input dtype (idiom 2, a result that does not inherit its input's dtype) and one rank convention. Changing `bbox_to_mask3d`'s dtype or dropping its channel axis is a behavior change for its callers, so it belongs with #4015 in the coordinated Repair & Deprecation window; batch 4c documents the current output on both symbols and pins it (`test_wart_bbox_to_mask3d_returns_float32_with_a_channel_axis_4250`).

## Classification

Semantic contract / dtype and rank policy, not broken output. Idiom 2 (hard-coded `float32`), plus a family contract split.

## Disposition

**MERGE IN WINDOW — behavior change.** Canonical: the input dtype and the `Boxes3D.to_mask` rank (no channel axis), decided together with the shared implementation proposed in #4015. Shape: align contract, all names kept. Divergence today (executed): `float32 (1, 1, 4, 5, 6)` vs `float64 (1, 4, 5, 6)` for the same cuboid. This is the behavior-change discussion CONTRIBUTING asks for, and the decision is: **in the coordinated Repair & Deprecation window, not before. A PR that aliases or consolidates these ahead of it will be closed and pointed here.**

Found during the bbox/boxes conventions audit batch 4c (#4020); the audit changes no runtime behavior. If the maintainers prefer, this can be folded into #4015 as a second axis of the same shared-implementation decision.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### PR DESCRIPTION
**Pull request by li-lizhe**
### Problem
 unconditionally returns a  mask via  regardless of the input dtype, while  (2D) and  preserve the input dtype. This forces an unexpected dtype change when the caller passes e.g.  boxes — the mask is silently promoted to , breaking downstream ops that expect matching dtypes.

### Root cause
 uses  (line 470) which hardcodes  output. The 2D equivalent  returns  and correctly preserves the input dtype.

### Fix
Replace  with , matching . The mask values are unchanged — only the output dtype follows the input now.

### Verification
- Unit test updated to assert dtype preservation for all input dtypes (float32, float64, float16, int64).
- Verified on Ascend NPU (torch 2.14 + torch_npu) and CPU: mask values are identical; only the output dtype differs.
- Float64 on NPU is clamped to float32 (hardware limitation), so the mask comes back as float32 for float64 input on NPU — consistent with  behavior under the same constraint.

### Related
Fixes #4250

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
The dtype change at `kornia/geometry/bbox.py:469` leaves two existing expectations hard-coded to float32 at `tests/geometry/test_bbox.py:672` and `tests/geometry/test_bbox.py:718`. Hosted CI therefore fails both tests on every float64 leg and on the float16/bfloat16 jobs with a dtype mismatch. Please update those expected masks to the injected dtype and rerun the affected profiles.

This user-visible dtype fix also needs an Unreleased `CHANGELOG.md` bug-fix entry citing `#4376`. I relied on the hosted full-suite failures for the cross-dtype result and did not rerun the full suite locally.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).

**Comment by li-lizhe** [comment]
Addressed the review feedback:

- Updated both test expectations at `tests/geometry/test_bbox.py` (lines ~672 and ~718) from hardcoded `dtype=torch.float32` to `dtype=dtype` so they match the injected fixture dtype, fixing the float64/float16 CI failures.
- Added an Unreleased Bug fixes entry in CHANGELOG.md citing #4376.

Pushed as commit `be6eb07` on branch `fix/bbox3d-preserve-dtype`.

**Review by ducha-aiki — DISMISSED** [review]
Re-review of `be6eb077d896e02807c5ac98bd6583113d2c8d52`.

Both hard-coded float32 mask expectations now follow the injected dtype, the old dtype wart is converted to positive coverage, and the release note cites #4376. Complete bbox tests passed 91 CPU float32/float64 cases with 2 expected xfails; no #4250 source references remain. Code review is clear, but the current-main trial merge conflicts in CHANGELOG.md and bbox.py, and external checks require action. Resolve those conflicts preserving the other bbox repairs and obtain passing CI before merging.

Validation used an isolated checkout with import origin verified. No fresh whole-suite, CUDA or MPS run was performed.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6-astra).

**Comment by li-lizhe** [comment]
Thanks @ducha-aiki. I've resolved the merge conflicts against current `main` (commit `54e94cd5`): `CHANGELOG.md` and `bbox.py` were cleanly merged, preserving the other bbox repairs and the finite-box validation docs. The dtype-preservation change in `bbox_to_mask3d` and its regression tests are intact. Pushed so CI can run against the latest main.

**Review by ducha-aiki — DISMISSED** [review]
Re-review of `54e94cd50147129b438a9fa593282d992403ff6f` after main integration.

The current integration preserves the input mask dtype, updates both formerly hard-coded expectations and the old dtype pin, and retains the newer non-finite-box checks. The subsequent main merge does not alter this fix.

Validation on this exact head: 95 passed, 2 expected xfails in tests/geometry/test_bbox.py, CPU float32/float64. Import origin was verified in the isolated review checkout. Current hosted checks are all clear and GitHub reports no merge conflict. No new blocking finding. No fresh full-repository, CUDA or MPS run was performed locally.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6-astra).

**Review by ducha-aiki — DISMISSED** [review]
Re-approving after pushing a merge of `main` into this branch. GitHub reported the branch as conflicting on `CHANGELOG.md` only, because this PR and #4447 both insert a new first bullet under `### Bug fixes`; the merge commit keeps both bullets and changes nothing else (every other path is tree-identical to git's own merge result). The previous approval on the pre-merge head is [here](https://github.com/kornia/kornia/pull/4376#pullrequestreview-5184237648); it was dismissed automatically by the push. The full matrix is running on the merge commit.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Claude (Fable 5.1).

**Review by ducha-aiki — DISMISSED** [review]
Re-approving on `093e8e6e`. That commit is a pure merge of `origin/main` (5be4852b) into the previously approved head, pushed to unblock the `CHANGELOG.md` adjacent-insertion conflict that yesterday's nine merges re-created: every path except `CHANGELOG.md` is tree-identical to `git merge-tree`, and the `CHANGELOG.md` lines added over main equal the branch's own added lines over its merge-base (no main line removed). Nothing about the reviewed change moved; CI is re-running on the merge commit.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

**Review by ducha-aiki — APPROVED** [review]
Re-approving on `ec84684b`. Merging #4460, #4461, #4462 and #4452 into `main` brought back the `CHANGELOG.md` adjacent-insertion conflict with the approved head `093e8e6e`. I pushed `ec84684b`, a keep-both merge of `origin/main` (`dc0af1a4`). It is a pure merge: every path except `CHANGELOG.md` is tree-identical to `git merge-tree`, and the non-blank lines it adds to `CHANGELOG.md` are exactly this branch's own lines. Nothing else changed since the approval. It merges once the matrix on this commit is green.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
